src/main/java/com/google/sps/servlets/GetBlurAreasServlet.java

```diff
@@ -0,0 +1,211 @@
+        for (FaceAnnotation face : res.getFaceAnnotationsList()) {
+          ArrayList<Point> facePoly = new ArrayList<>();
+          for (Vertex vertex : face.getFdBoundingPoly().getVerticesList())
+            facePoly.add(new Point(vertex.getX(), vertex.getY()));
+
+          // Add the rectangle to our list of face rectangles.
+          facePolys.add(facePoly);
+        }
+      }
+
+      client.close();
+    }
+
+    return facePolys;
+  }
+
+  /**
+   * Returns the BlobKey that points to the file uploaded by the user, or null if the user didn't
+   * upload a file.
+   */
+  private BlobKey getBlobKey(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a BlobKey. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so the BlobKey is empty. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    return blobKey;
+  }
+
+  /**
+   * Blobstore stores files as binary data. This function retrieves the image represented by the
+   * binary data stored at the BlobKey parameter.
+   */
+  private byte[] getBlobBytes(BlobKey blobKey) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
+      // End index is inclusive, so we have to subtract 1 to get fetchSize bytes.
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+
+      // If we read fewer bytes than we requested, then we reached the end.
+      if (b.length < fetchSize) {
+        continueReading = false;
+      }
+
+      currentByteIndex += fetchSize;
+    }
+
+    return outputBytes.toByteArray();
+  }
+
+  private Boolean isImageSupported(BlobKey blobKey) {
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    return supportedTypes.contains(blobInfo.getContentType());
+  }
+
+  /** Deletes a file from the blobstore */
+  private void deleteFile(BlobKey blobKey) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    blobstoreService.delete(blobKey);
```

Comment:
In some cases functions like this could throw an [exception](https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/blobstore/BlobstoreService.html#delete-com.google.appengine.api.blobstore.BlobKey...-), so you may want to wrap it up in a try-catch block and log and error message. The logging mechanism is actually something to think about as we want to know what's happening with the application. Maybe you could add logging as a separate PR - it will help you check that the application works as expected by logging 'success' info after each step, and if it doesn't work you'll see it in the error logs.

Reply:
Should I wrap the whole doPost method in a try-catch block? Blobstore service calls are mostly everywhere.

I think the logging system is a good idea, we could try to create one.

Reply:
Wrapping the whole doPost method sounds good to me, as well as logging caught exceptions to stdout. We can improve/extend logging in a separate PR.

Reply:
Wrapping the whole method it's not needed, all the exceptions inside the doPost method are automatically logged by maven.